elixir/apps/api/lib/api/client/channel.ex

```diff
@@ -507,6 +552,90 @@ defmodule API.Client.Channel do
     end
   end
 
+  # This message is sent to the client to request a network flow with a gateway that can server given resource.
+  #
+  # `connected_gateway_ids` is used to indicate that the client is already connected to some of the gateways,
+  # so the gateway can be reused by multiplexing the connection.
+  #
+  # `client_payload` is an arbitrary data that the client wants to send to the gateway, typically it will be
+  # either empty, or contain dns hostname of a given resource to connect to (so that it can be resolved on the
+  # gateway).
+  def handle_in(
+        "create_flow",
+        %{
+          "resource_id" => resource_id,
+          "connected_gateway_ids" => connected_gateway_ids
+        } = attrs,
+        socket
+      ) do
+    OpenTelemetry.Ctx.attach(socket.assigns.opentelemetry_ctx)
+    OpenTelemetry.Tracer.set_current_span(socket.assigns.opentelemetry_span_ctx)
+
+    OpenTelemetry.Tracer.with_span "client.create_flow", attributes: attrs do
+      location = {
+        socket.assigns.client.last_seen_remote_ip_location_lat,
+        socket.assigns.client.last_seen_remote_ip_location_lon
+      }
+
+      with {:ok, resource} <-
+             Resources.fetch_and_authorize_resource_by_id(resource_id, socket.assigns.subject),
+           {:ok, gateways} when gateways != [] <-
+             Gateways.all_connected_gateways_for_resource(resource, socket.assigns.subject,
+               preload: :group
+             ),
+           {:ok, gateways} <-
+             filter_compatible_gateways(gateways, socket.assigns.gateway_version_requirement),
+           OpenTelemetry.Tracer.set_attribute(:gateways_count, length(gateways)),
+           gateway = Gateways.load_balance_gateways(location, gateways, connected_gateway_ids),
+           OpenTelemetry.Tracer.set_attribute(:gateway_id, gateway.id),
+           {:ok, resource, flow} <-
+             Flows.authorize_flow(
+               socket.assigns.client,
+               gateway,
+               resource_id,
+               socket.assigns.subject
+             ) do
+        OpenTelemetry.Tracer.set_attribute(:flow_id, flow.id)
+        opentelemetry_ctx = OpenTelemetry.Ctx.get_current()
+        opentelemetry_span_ctx = OpenTelemetry.Tracer.current_span_ctx()
+
+        preshared_key = Domain.Crypto.psk()
```

Comment:
@thomaseizinger do we need PSK to be unique every time or should I make one per client session? If it's one per session what happens when WebSocket reconnects?

Reply:
We will only use it when we make a new connection, i.e. when the ICE credentials change. I think to avoid confusion, it would be good if it would change at the same cadence as the ICE credentials.

Reply:
Hm, we could just reuse the same hash we use for ICE credentials but I'm not sure that will be cryptographically correct way to do it - the portal does not control the pubkeys generation so it might not have enough entropy to be used as a seed for preshared key (vs using strong random bytes from OS).